### PR TITLE
Implement CLI logging

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
## Description

Adds the necessary auditing outputs defined by the Wiki file: https://github.com/Azure/platform-chaos/wiki/Auditing

## Checklist for success

- [x] This change isn't already implemented by a pending pull request
- [x] This change is a child commit of a reasonably recent commit in the **master** branch 
    * Requests need not be a single commit, but should be a linear sequence of commits (i.e. no merge commits in your PR)
- [x] All tests pass locally (see [the testing section](https://github.com/azure/platform-chaos-cli/blob/master/README.md#testing) for more info)
- [x] All New behaviors or major functional changes have tests

### CODE REVIEW

File: lib/request-processor.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const request = require('request')
 const os = require('os')
 const uuidv4 = require('uuid/v4')
 const logger = require('./logger')
-
+const assert = require('assert')
```
Comment (bengreenier) [inline lib/request-processor.js]: nit: ordering/spacing.

should come before `logger`, should have a newline after require calls

File: lib/request-processor.js [hunk 2]
```diff
@@ -22,6 +22,9 @@ module.exports = class RequestProcessor {
   }
 
   start ({extensionUri, accessToken, resources, authKey}) {
+    if (extensionUri === null && typeof extensionUri !== 'string') {
```
Comment (bengreenier) [inline lib/request-processor.js]: let's drop null check, as discussed

File: lib/request-processor.js [hunk 3]
```diff
@@ -51,6 +54,7 @@ module.exports = class RequestProcessor {
   }
 
   stop ({extensionUri, accessToken, resources, authKey}) {
+    assert.ok(extensionUri !== null && typeof extensionUri === 'string')
```
Comment (bengreenier) [inline lib/request-processor.js]: let's drop null check as discussed

File: .vscode/launch.json [hunk 4]
```diff
@@ -1,31 +0,0 @@
-{
```
Comment (bengreenier) [inline .vscode/launch.json]: why did we delete this?

File: bin/chaos.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ const home = require('home')
 const yargs = require('yargs')
 const factory = require('../lib/factory')
 const pkgJson = require('../package.json')
-
+const logger = require('../lib/logger')
```
Comment (bengreenier) [inline bin/chaos.js]: nit: newline after requires

File: lib/factory.js [hunk 6]
```diff
@@ -1,9 +1,8 @@
 const RequestProcessor = require('./request-processor')
 const ExtensionRegistry = require('./extension-registry')
 const AzureAuthenticator = require('./azure-authenticator')
-const Logger = require('./logger')
 
-const instanceGenerator = (Type) => {
+const instanceGenerator = Type => {
```
Comment (bengreenier) [inline lib/factory.js]: nit: why did we make this change?

File: lib/logger.js [hunk 7]
```diff
@@ -1,32 +1,38 @@
...
+  }
+
+  configure ({ logImpl }) {
+    this._logImpl = logImpl
   }
 
   info (...params) {
     params.unshift('[INFO]')
-    this._logImpl(this, chalk.white.bgBlue.apply(this, params))
+    this._logImpl(chalk.white.bgBlue(...params))
```
Comment (bengreenier) [inline lib/logger.js]: nice, these are much cleaner

File: lib/request-processor.js [hunk 8]
```diff
@@ -1,11 +1,30 @@
 const request = require('request')
-
+const os = require('os')
+const uuidv4 = require('uuid/v4')
+const logger = require('./logger')
+const assert = require('assert')
 module.exports = class RequestProcessor {
```
Comment (bengreenier) [inline lib/request-processor.js]: nit: newline after requires

File: lib/request-processor.js [hunk 9]
```diff
@@ -1,11 +1,30 @@
...
   constructor ({requestImpl}) {
     this._request = requestImpl || request
+    this._logger = logger
+  }
+
+  _log ({extensionUri, accessToken, resources, type}) {
+    this._logger.log(JSON.stringify({
+      'id': uuidv4(),
+      'type': type,
+      'user': accessToken.substr(7, 16), // get the first 16 characters of the user's access token
```
Comment (bengreenier) [inline lib/request-processor.js]: this seems oddly specific and arbitrary?

File: lib/request-processor.js [hunk 10]
```diff
@@ -1,11 +1,30 @@
...
+      'system': `${os.hostname()}-${os.type()}`,
+      'date': new Date(),
```
Comment (bengreenier) [inline lib/request-processor.js]: how does this get serialized by default?

File: test/azure-authenticator.js [hunk 11]
```diff
@@ -1,9 +1,10 @@
 const assert = require('assert')
 const factory = require('../').factory
-
+const logger = require('../lib/logger')
 /* eslint-env node, mocha */
 
 describe('azure-chaos', () => {
+  beforeEach(() => logger.configure({ logImpl: () => null })) // noop by default
```
Comment (bengreenier) [inline test/azure-authenticator.js]: nit: newline after

File: test/azure-authenticator.js [hunk 12]
```diff
@@ -77,8 +78,8 @@ describe('azure-chaos', () => {
         resources: testObject.expectedResources,
         authKey: testObject.expectedAuthKey,
         accessToken: testObject.expectedAccessToken
-      }).then((res) => {
-        assert.throws(res.calledUri, null)
+      }).then(null, err => {
```
Comment (bengreenier) [inline test/azure-authenticator.js]: nit: inconsistent styling - move the next `then` to the same line as closers, or move this `then` to it's own line

File: test/azure-authenticator.js (hunk 11 above)
Comment (bengreenier) [reply test/azure-authenticator.js]: further, this is inconsistent, please standardize to either all one line, or multi line

File: test/request-processor.js [hunk 13]
```diff
@@ -1,9 +1,12 @@
 const assert = require('assert')
 const factory = require('../').factory
-
+const logger = require('../lib/logger')
 /* eslint-env node, mocha */
 
 describe('RequestProcessor', () => {
+  beforeEach(() => {
```
Comment (bengreenier) [inline test/request-processor.js]: see comment above, please standardize

File: test/request-processor.js [hunk 14]
```diff
@@ -86,4 +89,38 @@ describe('RequestProcessor', () => {
...
+          const response = { calledUri, options }
+          return cb(null, response)
+        }
+      }
+    })
+
+    logger.configure({
+      logImpl: (message) => {
+        assert.ok(typeof message === 'string')
+        assert.ok(/\[LOG\]/.test(message))
```
Comment (bengreenier) [inline test/request-processor.js]: how do we assert this code even runs? ie: right now, if nothing is ever logged, this test passes.

File: .vscode/launch.json (hunk 4 above)
Comment (Ethan-Arrowood) [reply .vscode/launch.json]: restoring in next commit

File: lib/request-processor.js (hunk 9 above)
Comment (Ethan-Arrowood) [reply lib/request-processor.js]: The full accessToken is very long and clutters the log. I find these first 16 characters would be good enough to determine which authenticated user executed the chaos event.

Thinking back to why this prop was added in the first place: The audit should give some identifier as to _who_ executed the event. If this is a person or some automated system; everything requires some accessToken in order to emit chaos.

Let me know your thoughts on what a good resolution would be here.

File: lib/request-processor.js (hunk 10 above)
Comment (Ethan-Arrowood) [reply lib/request-processor.js]: nice catch. new Date() auto serializes to the ISO string. I'm adding `.toISOstring()` in order to make this more clear.

File: test/request-processor.js (hunk 14 above)
Comment (Ethan-Arrowood) [reply test/request-processor.js]: Good point. I'll see if I can set up a spy to ensure the method is running.

File: lib/request-processor.js [hunk 15]
```diff
@@ -1,11 +1,29 @@
 const request = require('request')
-
+const os = require('os')
+const uuidv4 = require('uuid/v4')
+const logger = require('./logger')
```
Comment (allydurks) [inline lib/request-processor.js]: nit: newline after the requires

File: package.json [hunk 16]
```diff
@@ -11,7 +11,9 @@
     "home": "^1.0.1",
     "ms-rest-azure": "^2.5.5",
     "request": "^2.85.0",
+    "sinon": "^6.2.0",
```
Comment (allydurks) [inline package.json]: shouldn't this be a dev-dependency if you're only using it in the tests?

**Pull request merged**